### PR TITLE
Break week poc: improve testability and introduce a clearer code structure

### PR DESCRIPTION
**Pull request by vsupalov**
Based on the idea that we can make the code core more testable by moving client interactions out of the controller business logic.

* introduce three steps to the controller loop
* add data structure to hold additional data
* create data structure to hold cluster resources to be applied
* rework internal secret creating into new steps
* add some representative test cases on generated YAML

# Description

*Please add a description here. This will become the commit message of the merge request later.*

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by vsupalov** [comment]
Highlights of this POC PR:

* the code structure is nicer, it feels pretty clean
* all new-style tests run in 0.2 seconds. The implications for the developer experience are immense.
* we have a smoke test for the yaml core of the application (smoke test).
* a kuttl test could be replaced with a [rust-native test](https://github.com/stackabletech/druid-operator/pull/391/files#diff-ccce1b4b66a0ab04cb82b37ee803f69ee8bf905f6c836e0ba9216a06998d1484R287) (see kuttl/resources).
* previously untested funcationality (config overrides) have not been tested previously due to overhead. [It is now](https://github.com/stackabletech/druid-operator/pull/391/files#diff-ccce1b4b66a0ab04cb82b37ee803f69ee8bf905f6c836e0ba9216a06998d1484R241).

Downside: we have to assume that the generated YAMLs will be accepted by k8s. This is not tested.

File: rust/operator-binary/src/controller/apply.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        source: stackable_operator::error::Error,
+    },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn handle_cluster_resources(
+    client: &Client,
+    druid: &Arc<DruidCluster>,
+    appliable_cluster_resources: Vec<AppliableClusterResource>,
```
Comment (razvan) [inline rust/operator-binary/src/controller/apply.rs:69]: I think you can do away with `AppliableClusterResource` and just use `ClusterResource`:
```suggestion
    appliable_cluster_resources: Vec<ClusterResource>,
```
Then this function just loops over the array and applies each element. And then it's just a generic function in operator-rs. It's even future-proof if we add more `ClusterResource` types.

File: rust/operator-binary/src/controller/apply.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            AppliableClusterResource::RolegroupStatefulSet(rg_statefulset, rolegroup) => {
+                cluster_resources
+                    .add(client, &rg_statefulset)
+                    .await
+                    .with_context(|_| ApplyRoleGroupStatefulSetSnafu {
+                        rolegroup: rolegroup.clone(),
+                    })?;
+            }
+            AppliableClusterResource::InternalSecret(secret) => {
+                // TODO: I must admit, I don't quite know what's going on here right now
```
Comment (razvan) [inline rust/operator-binary/src/controller/apply.rs:118]: Looks like leftovers from the times when `Secret` was not a `ClusterResource`

File: rust/operator-binary/src/controller/build.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
+mod config;
```
Comment (razvan) [inline rust/operator-binary/src/controller/build.rs:1]: I always thought [build.rs](https://doc.rust-lang.org/cargo/reference/build-scripts.html) is a bit special for cargo.

File: rust/operator-binary/src/controller/build.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    },
+    #[snafu(display("failed to retrieve secret for internal communications"))]
+    FailedInternalSecretCreation { source: internal_secret::Error },
+    #[snafu(display("failed to build role related cluster resource"))]
+    RoleBuildError { source: roles::Error },
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub async fn create_appliable_cluster_resources(
```
Comment (razvan) [inline rust/operator-binary/src/controller/build.rs:64]: As far as I see, this function doesn't need to be `async` (takes no `client`)

**Review by razvan — COMMENTED** [review]
I really wish we would use more traits and less enums.

File: rust/operator-binary/src/controller/build.rs (hunk 3 above)
Comment (vsupalov) [reply rust/operator-binary/src/controller/build.rs:1]: Only in the root of a package.

File: rust/operator-binary/src/controller/apply.rs (hunk 1 above)
Comment (vsupalov) [reply rust/operator-binary/src/controller/apply.rs:69]: There are upsides and downsides to both choices. I went with enums to easily pass additional information along with the `ClusterResource` and keep scope of necessary code changes at repo level.

**Comment by vsupalov** [comment]
Thanks for taking a look @razvan!

File: rust/operator-binary/src/controller/build.rs (hunk 4 above)
Comment (vsupalov) [reply rust/operator-binary/src/controller/build.rs]: Good point! I took this for granted. Couldn't resist fixing this.

**Comment by fhennig** [comment]
I think I like this, how are we going to proceed with this? If we don't merge it, it will go stale, but maybe that's fine?

**Comment by razvan** [comment]
will not implement

**Pull request closed without merging**